### ISSUE TITLE
Get detailed sales and receiving data in the detailed reports by ajax 

### ISSUE THREAD

**Issue by freemail624** [jekkos/opensourcepos#1227]
Instead of loading all the data for each detail view we can pull it using ajax only when the row is shown.
i taked the idea from here [#953](https://github.com/wenzhixin/bootstrap-table/issues/953)
I tried this solution and worked with me
In the tabular_details we can replace this code
`var detail_data = <?php echo json_encode($details_data); ?>;`
`$detail.html('<table></table>').find("table").bootstrapTable({
					columns: <?php echo transform_headers_readonly($headers['details']); ?>,
					data: detail_data[(!isNaN(row.id) && row.id) || $(row[0] || row.id).text().replace(/(POS|RECV)\s*/g, '')]
				});`
with this code
assigned the ajax link to a variable 
`var detail_data = <?php echo "'".$ajax_data_link ."'"; ?>; // ex: 'sales/get_sale_items/'`
`$.get(detail_data+row.id, function (res) {
					$detail.html('<table></table>').find("table").bootstrapTable({
						columns: <?php echo transform_headers_readonly($headers['details']); ?>,
						data: JSON.parse(res),
					});
				});`

And in the controller we will create function to get the items 
like this for sales

public function get_sale_items($sale_id)
	{
		$sale_items = $this->Sale->get_sale_items($sale_id)->result_array();
		foreach ($sale_items as $row)
		{
			$data[]=$this->xss_clean(array($row['item_id'], $row['quantity_purchased'], to_currency_no_money($row['item_unit_price']), to_currency_no_money($row['discount_percent']), to_currency_no_money(($row['item_unit_price']*$row['quantity_purchased'])-($row['item_unit_price']*$row['quantity_purchased']*$row['discount_percent']/100))));
		}
		echo json_encode($data);
	}

I wish this will work with you

**Comment by daN4cat** [comment]
As far as I remember only the page data is loaded and not everything.
What you are suggesting implies a lot of back and forth between the front end and the webserver/backend.
You benefit less of any caching and load the system for no real gain.

Unless you managed to enable an infinite scrolling of the UI tabular view so you really load what on display. But I don't seem to see it from your suggestion.

Or am I missing the point here?

**Comment by freemail624** [comment]
So why the detailed reports very slow than the summary reports 
when i try to get the data for one year the summary reports get the data faster than the detailed reports 
it take long time and make the program very slow
I think that the detailed reports take long time because the query select all the sales and the sales items at the same time, so i want to make it more faster by generate only the sales data and get the sales items when needed by ajax

**Comment by daN4cat** [comment]
Which version are you talking about?

**Comment by freemail624** [comment]
version 3.0.0

**Comment by daN4cat** [comment]
Ok, old stuff. I made huge improvements with later versions.

**Comment by daN4cat** [comment]
I mean about performance of database loading.

**Comment by daN4cat** [comment]
Please test with the latest code in master and see if the issue is still there.

**Comment by freemail624** [comment]
The way of all versions for the detailed reports is to select the summary and the detailed data at the same time.
Suppose we have 100000 sales receipts and 500000 sales items .
If we generate a detailed sales report for these sales receipts, it will select all the 100000 sales receipts and the 500000 sales items at the same time.
In the report view it show only the 100000 sales receipts divided on pages and if you want the sales items you can click on the toggle plus to show them.

(The point here is that the detailed data not showed until the user need to show it manually)

My view is instead of making the query selecting 600000 records (the summary and the detailed data) 
we can make it selecting only 100000 records (the summary data)  and if you want the detailed data you can click on the toggle plus to show them by ajax.
So the query which select 100000 records is more faster than the query which select 600000 records.
I wish that my view is correct.

**Comment by daN4cat** [comment]
@freemail624 yes got your point. I checked the code in the models because I couldn't remember exactly anymore.
I remember that the paging point was discuss with @jekkos last year when he reworked the tables during the bootstrapisation of OSPOS.

What I've done at the time was to narrow down all the searches and temp table creation to the selected time range. It used to create temp tables with everything in it!
Your suggestion means that for the summary (totals) you still need to use what it is now with no page limit and then add the option to fetch just the data for the page range even changing the temp table.
I'm not sure how much you would gain on performance changing the table everytime. That would surely needs performance testing. For sure the amount of memory used by MySQL for those temp tables will reduce.

Check `application/models/reports/Summary_report.php private function _common_select(array $inputs)` as starting point, you would need to add a page number and limit in the inputs. If I remember right the controller works in that way. Maybe takings already work in that way, I can't really remember.

If you want to get logs of database latencies we added a db log feature, just go to `application/config/config.php` and set `$config['db_log_enabled']` to TRUE. Then in `application/logs` you should find logs of the db calls with timing.

**Comment by daN4cat** [comment]
OK, maybe I sidetracked here.... you probably just meant to work on the json payload between frontend and web server, right?

**Comment by daN4cat** [comment]
Your suggestion could be really another optimisation.

**Comment by jekkos** [comment]
Yep indeed you're right all details are directly written to the page once it's loaded now. . I guess one would need to move detail loading to a separate function and call it by sale_id for one row or pass it the same criteria as in the main screen to load all rows at the same time.
